pypuf/property_test/base.py

```diff
@@ -0,0 +1,172 @@
+        """
+        reliabilities = []
+        n = len(challenges[0])
+        for ins in instances:
+            for challenge in challenges:
+                shaped_challenge = reshape(challenge, (1, n))
+                reliabilities.append(PropertyTest.reliability(ins, shaped_challenge, measurements=measurements))
+        return reliabilities
+
+    def reliability_statistic(self, challenges, measurements=10, alpha=0.05):
+        """
+        This function calculates the reliability statistic.
+        :param challenges: array of int shape(N,n)
+        :param measurements: int default 10
+                             Number of evaluations of the puf instance.
+        :param alpha: float 0 < alpha < 1
+                      Error level
+        :return: Dictionary of float
+                 Statistic of the set of instances {mean:, median:, min:, max:, sample_variance:, confidence_int:}
+        """
+        return PropertyTest.statistic(
+            PropertyTest.reliability_set(self.instances, challenges, measurements=measurements),
+            alpha=alpha,
+        )
+
+    @staticmethod
+    def uniqueness(instances, challenge):
+        """
+        This function calculates the uniqueness of a challenge response for a set of simulation instances.
+        :param instances: array of pypuf.simulation.base.Simulation with shape(k)
+        :param challenge: challenge: array of int shape(1,n)
+        :return: float
+                 Uniqueness in percent.
+        """
+        responses = array([instance.eval(challenge) for instance in instances])
+        m = len(instances)
+        distance_sum = 0
+        for u in range(m - 1):
+            for v in range(u + 1, m):
+                # If we get simulations with response length > 1 then change the calculation of the distance.
+                distance_sum = distance_sum + (responses[u] != responses[v])
+        # Arithmetic mean of the sum of response distances
+        return 2 / (m * (m - 1)) * distance_sum * 100
+
+    @staticmethod
+    def uniqueness_set(instances, challenges, measurements=1):
+        """
+        This function calculates a uniqueness set for a list of instances, challenges and measurements.
+        :param instances: array of pypuf.simulation.base.Simulation with shape(k)
+        :param challenges: challenge: array of int shape(N,n)
+        :param measurements: int default 10
+                             Number of uniqueness calculation for fix arguments.
+        :return: list of float
+                 List of uniqueness.
+        """
+        n = len((challenges[0]))
+        uniqueness_set = []
+        for challenge in challenges:
+            for _ in range(measurements):
+                uniqueness_set.append(PropertyTest.uniqueness(instances, reshape(challenge, (1, n))))
+        return uniqueness_set
+
+    def uniqueness_statistic(self, challenges, measurements=10, alpha=0.05):
```

Comment:
this function mostly doubles functionality that we have already seen in `reliability_statistics`. Maybe it makes sense to merge those two functions?

Reply:
You are right that the parameter of `uniqueness_set` and `reliability_set` are the same. But the function bodies differ from each other. 
```
   @staticmethod
    def uniqueness_set(instances, challenges, measurements=1):
        """
        This function calculates a uniqueness set for a list of instances, challenges and measurements.
        :param instances: array of pypuf.simulation.base.Simulation with shape(k)
        :param challenges: challenge: array of int shape(N,n)
        :param measurements: int default 10
                             Number of uniqueness calculation for fix arguments.
        :return: list of float
                 List of uniqueness.
        """
        n = len((challenges[0]))
        uniqueness_set = []
        for challenge in challenges:
            for _ in range(measurements):
                uniqueness_set.append(PropertyTest.uniqueness(instances, reshape(challenge, (1, n))))
        return uniqueness_set

    @staticmethod
    def reliability_set(instances, challenges, measurements=10):
        """
        This function calculates a set of reliabilities.
        :param instances: instances: list of pypuf.simulation.base.Simulation
        :param challenges: array of int shape(N,n)
        :param measurements: int default 10
                             Number of evaluations of the puf instance.
        :return: list of float
                 Array of reliabilities for the puf instances, challenges and measurements.
        """
        reliabilities = []
        n = len(challenges[0])
        for ins in instances:
            for challenge in challenges:
                shaped_challenge = reshape(challenge, (1, n))
                reliabilities.append(PropertyTest.reliability(ins, shaped_challenge, measurements=measurements))
        return reliabilities
```
As I understand you correctly, you want to have a less redundant realization like

```
def float_set(property_function, instance, challenges, measurements):
    for challenge in challenges:
            for _ in range(measurements):
                property_function(instances, challenge)
```

I focused on keeping my implementation near to the formulas found in my MT and literature to increase the transparency of my work and to gain the credibility. On top of that, I like to call interchange the function, not the parameter as we did for `LTFArray.transformation` and `LTFArray.combiner`.